### PR TITLE
Add Data types section

### PR DESCRIPTION
**Pull request by mmaelicke**

### CODE REVIEW

File: source/developer/wps.rst [hunk 1]
```diff
@@ -153,28 +153,30 @@ The definition of inputs follows ``LiteralInput(identifier, title=None, data_typ
...
     - multi select dropdown
 
 * allowed_values - defined values exposed to the user as **radio buttons**.
 * default - value used when the user selects nothing for this input.
 
 The definition of the file-based data types is supposed to be done in ``Keywords``, when they are implemented in ``pywps`` and ``owslib``.
-From the example above, ``min_occurence=2`` forces a ``*required`` flag to the input field, and ``max_occurence=5`` forces a multiple select option for the portal user.
+
+The example above results in a dropdown that shows ``'UUID of the timeseries to be used'`` when the user hovers over the element. The user can select two to five timeseries data-sets from the dropdown (``min_occurence=2`` forces a ``*required`` flag to the input field, and ``max_occurence=5`` forces a multiple select option).
+
```
Comment (mmaelicke) [inline source/developer/wps.rst]: I love this approach! This gonna be awesome.

File: source/developer/index.rst [hunk 2]
```diff
@@ -1,15 +1,16 @@
...
+
+    Define sub-pages and add to toctree below
+
+.. toctree::
+    :maxdepth: 2
+    :caption: Contents:
+
+    install 
+    standalone
+    wps
```
Comment (MarcusStrobl) [inline source/developer/index.rst:15]: Wonderful

File: source/developer/wps.rst (hunk 1 above)
Comment (MarcusStrobl) [reply source/developer/wps.rst]: Yes! :)
Thank you!

**Review by MarcusStrobl — APPROVED** [review]
Well done.

**Comment by mmaelicke** [comment]
Thanks @MarcusStrobl! 

@sihassl, any typos or suggestions left?

File: source/developer/wps.rst [hunk 3]
```diff
@@ -0,0 +1,288 @@
+===========
+WPS Toolbox
+===========
+
+Overview
+========
+
+The Tools in V-FOR-WaTer are implemented as a standalone server. It uses the `WPS Protocol <https://de.wikipedia.org/wiki/Web_Processing_Service>`_ to receive
+inputs, data and configuration and returns processing output. Most data, including input data from the database, but also processing results and intermediate results
```
Comment (sihassl) [inline source/developer/wps.rst]: inputs, data and configuration and returns processing output. Most data, including input data from the database, but also processing results and intermediate results,

File: source/developer/wps.rst [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+are saved to a temporary folder on the processing server. The WPS just exchanges configuration and file-locations. For this to work, a set of *data types* are defined.
+Each tool can only handle a specific, or a set of specified data types and has defined outputs. These data-types are hierachical. That means there are tools that
```
Comment (sihassl) [inline source/developer/wps.rst]: Each tool can only handle one specific data type, or a set of specified data types, and has defined outputs. These data types are hierachical. That means there are tools that

File: source/developer/wps.rst [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+    - idataframe
+      - vdataframe
+    - time-dataframe
+      - vtime-dataframe
+   html
+    - plot
+
+The hierachy is *downward* compatible. That means a named variable ``vtimeseries`` is also a valid ``timeseries`` and an ``array``, but not a `varray`. It is possbile that tools might accept data types from different branches. A tool that accepts ``array`` **and** ``ndarray`` will literally take *any* input.
+
+Defining Data-Types
```
Comment (sihassl) [inline source/developer/wps.rst]: _data types_

File: source/developer/wps.rst [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+    - ``>1``
+    - one or more values & required
+    - multi select dropdown
+
+* allowed_values - defined values exposed to the user as **radio buttons**.
+* default - value used when the user selects nothing for this input.
+
+The definition of the file-based data types is supposed to be done in ``Keywords``, when they are implemented in ``pywps`` and ``owslib``.
+
+The example above results in a dropdown that shows ``'UUID of the timeseries to be used'`` when the user hovers over the element. The user can select two to five timeseries data-sets from the dropdown (``min_occurence=2`` forces a ``*required`` flag to the input field, and ``max_occurence=5`` forces a multiple select option).
```
Comment (sihassl) [inline source/developer/wps.rst]: The example above results in a dropdown that shows ``'UUID of the timeseries to be used'`` when the user hovers over the element. The user can select two to five timeseries datasets from the dropdown (``min_occurence=2`` forces a ``*required`` flag to the input field, and ``max_occurence=5`` forces a multiple select option).

File: source/developer/wps.rst [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+
+  {
+    "error": True | False,
+    "message": "error message if any",
+    "type": "error type if any"
+  }
+
+The error-object will also be returned, if there is no error. Then, the ``error['error']==False``. The error type can be one of:
+
+* bug - unexpected Exceptions that are not handled in the Toolbox. These should be reported to the developer
```
Comment (sihassl) [inline source/developer/wps.rst]: * bug - unexpected exceptions that are not handled in the Toolbox. These should be reported to the developer.

File: source/developer/wps.rst [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+* userWarning - mainly due to wrong options passed.
+* processError - expected errors that are specific to the tool. These errors need to be reported to the user.
+
+An example of a processError would be a ``numpy.LinAlgError`` that is raised during Kriging, if the kriging matrix is bad conditioned.
```
Comment (sihassl) [inline source/developer/wps.rst:203]: An example of a processError would be a ``numpy.LinAlgError`` that is raised during Kriging, if the kriging matrix is badly conditioned.

File: source/developer/wps.rst [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+This is an expected error that should be reported to the user, as it could indicate that the Kriging results are incorrect.
+These kind of errors cannot be handled in the Toolbox.
```
Comment (sihassl) [inline source/developer/wps.rst:205]: These kinds of errors cannot be handled in the Toolbox.

File: source/developer/wps.rst [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+
+Metadata file
+-------------
+
+Each tool will also write a metadata file in ``.json`` format. This metadata file contains
+metadata about the initial datasets, that might be required by some tools and a collection of all tools that already were applied.
```
Comment (sihassl) [inline source/developer/wps.rst]: metadata about the initial datasets that might be required by some tools, and a collection of all tools that were already applied.

File: source/developer/wps.rst [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+That means, for a specific tool, the corresponding ``.json`` will conain the UUIDs of other tool runs.
+This way a toolchain can be traced. If you re-run a tool, new ``.json`` will be written.
```
Comment (sihassl) [inline source/developer/wps.rst]: In this way, a toolchain can be traced. If you re-run a tool, a new ``.json`` will be written.

File: source/developer/wps.rst [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+
+The ``.json`` file will contain all information about the tool processed. This includes metadata like the tool name and the
```
Comment (sihassl) [inline source/developer/wps.rst]: The ``.json`` file will contain all information about the processed tool. This includes metadata like the tool name and the

File: source/developer/wps.rst [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+inputs given, but also parameters like the processing time. For any input, that was the result of another tool, the UUID is
```
Comment (sihassl) [inline source/developer/wps.rst]: given inputs, but also parameters like the processing time. For any input which was the result of another tool, the UUID is

File: source/developer/wps.rst [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+
+  ec97a39e-933c-411f-b4f5-7716871bb64f.csv
+  ec97a39e-933c-411f-b4f5-7716871bb64f.json
+  6b23c7b6-4457-4ff2-82a3-e3c5136abfe0.script.html
+  6b23c7b6-4457-4ff2-82a3-e3c5136abfe0.div.html
+  6b23c7b6-4457-4ff2-82a3-e3c5136abfe0.json
+
+.. note::
+
+  At this point the two json files essentially contain the same information, but as the ``vtimeseries`` might be further processed.
```
Comment (sihassl) [inline source/developer/wps.rst]: At this point the two ``.json`` files essentially contain the same information, but as the ``vtimeseries`` might be further processed.

something missing in this sentence? or is the "as" too much?

**Comment by mmaelicke** [comment]
I took over all comments and will, therefore, merge without requesting new approval.

**Pull request merged**